### PR TITLE
Features: Tokenizer, basic trainer script.

### PR DESCRIPTION
**Pull request by h-alice**
In this PR, we implemented tokenizer and dataset-loader components.

### CODE REVIEW

File: requirements.txt
Comment (NatLee) [inline requirements.txt:1]: It's good to add version with packages 😀

File: src/core/llm_connector.py [hunk 1]
```diff
@@ -146,7 +149,7 @@ class LlmConnectorLlamacpp(LlmConnector):
     def __init__(self, model_name: str, model_path: str, verbose: bool = False, max_tokens: int = 2048):
 
         # Call the parent class constructor
-        super().__init__(model_provider=ModelProviders.LLAMACPP, model_name=model_name)
+        super().__init__(model_provider=str(ModelProviders.LLAMACPP), model_name=model_name)
```
Comment (NatLee) [inline src/core/llm_connector.py:152]: Why here needs forcing it to be a String?

File: src/trainer/tokenization.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        token_list["input_ids"].append(tokenizer.eos_token_id) # type: ignore
+        token_list["attention_mask"].append(1) # Set the attention mask to 1. # type: ignore
+
+    return token_list
+
+
+
+if __name__ == "__main__": # For test.
+    tokenizer = AutoTokenizer.from_pretrained("/Users/h-alice/Desktop/gemma-2b")
+    print(tokenize(tokenizer, "來去菜市場看歐郎甲追夠", eos_token=False)) # type: ignore
```
Comment (NatLee) [inline src/trainer/tokenization.py:36]: uwu

File: src/trainer/tokenization.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            text, # Prompt.
+            truncation=True, # Truncate the text if it exceeds the maximum length.
+            max_length=max_length, # Maximum length of the tokenized text.
+            padding=False, # No padding.
+        )
+    
+    if eos_token:
+        # Append the EOS token. 
+        # Note that it will append the token even if it exceeds the maximum length.
+        token_list["input_ids"].append(tokenizer.eos_token_id) # type: ignore
```
Comment (NatLee) [inline src/trainer/tokenization.py:27]: More keys!!! But official package 🫠

**Review by NatLee — APPROVED** [review]
It's good, but we need to add version to our packages in requirements. 🥹

File: src/trainer/tokenization.py (hunk 2 above)
Comment (h-alice) [reply src/trainer/tokenization.py:36]: uwu

File: src/core/llm_connector.py (hunk 1 above)
Comment (h-alice) [reply src/core/llm_connector.py:152]: @NatLee Because Pylance (the type checker in VSCode) will complain about implicit type casting between enums and string. That's annoying. : (

Comment (h-alice) [reply src/core/llm_connector.py:152]: So we need to add the string casting to avoid errors.

File: src/trainer/tokenization.py (hunk 3 above)
Comment (h-alice) [reply src/trainer/tokenization.py:27]: In this cast, the returned value is a dict which contains token index list (input_ids) and attention mask.
We followed the implementation of official documentation here.

File: requirements.txt
Comment (h-alice) [reply requirements.txt:1]: It is a good idea. But how if the dependent package rapidly changes their versions?

Comment (h-alice) [reply requirements.txt:1]: Fixed in commit https://github.com/h-alice/trashtalk-bot/pull/2/commits/423defda118505aaff977e4aa75b0af12c2cf7ba.
Thanks for suggestion!

**Comment by h-alice** [comment]
OK, let's GOOOOO!

File: requirements.txt
Comment (NatLee) [reply requirements.txt:1]: In my view, Git's `Tag` can be used to control the packages' version.
